modules/core/src/test/java/org/apache/ignite/internal/util/collection/AbstractBaseIntMapTest.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    /**
+     *
+     */
+    @Test
+    public void putIfAbsent() {
+        IntMap<String> map = instantiateMap();
+
+        String newVal = map.putIfAbsent(1, value(1));
+
+        assertNull(newVal);
+
+        assertEquals(value(1), map.get(1));
+
+        String retry = map.putIfAbsent(1, value(2));
+
+        assertEquals(value(1), retry);
+    }
+
+    /**
+     *
+     */
+    @Test
+    public void forEach() {
+        IntMap<String> map = instantiateMap();
+
+        for (int i = 1; i < 100_000; i++)
+            map.put(i, value(i));
+
+
+        final AtomicLong cntr = new AtomicLong(0);
+
+        map.forEach((key, value) -> cntr.addAndGet(key));
+
+        assertEquals(99_999L * 100_000L / 2, cntr.get());
+    }
+
+    /**
+     *
+     */
+    @Test
+    public void contains() {
+        IntMap<String> map = instantiateMap();
+
+        for (int i = 1; i < 10_000; i++)
+            map.put(i, value(i));
+
+        for (int i = 1; i < 10_000; i++) {
+            assertTrue(map.containsKey(i));
+            assertTrue(map.containsValue(value(i)));
+        }
+
+        assertFalse(map.containsKey(0));
+        assertFalse(map.containsValue(value(0)));
+    }
+
+    /**
+     *
+     */
+    @Test
+    public void compareWithReferenceImplementation() {
+        Map<Integer, String> originMap = new HashMap<>();
+        IntMap<String> testable = instantiateMap();
+
+        ThreadLocalRandom randomGen = ThreadLocalRandom.current();
+
+        for (int i = 0; i < 10_000_000; i++) {
+            int nextKey = randomGen.nextInt(0, 1_000_000);
+            int actId = randomGen.nextInt(0, 4);
+
+            if (actId == 0) {
+                String oPut = originMap.put(nextKey, value(nextKey));
+                String ePut = testable.put(nextKey, value(nextKey));
+
+                assertEquals(oPut, ePut);
+                assertEquals(originMap.containsKey(nextKey), testable.containsKey(nextKey));
+            }
+            else if (actId == 1) {
+                assertEquals(originMap.get(nextKey), testable.get(nextKey));
+                assertEquals(originMap.containsKey(nextKey), testable.containsKey(nextKey));
+            }
+            else if (actId == 2) {
+                String oPutAbs = originMap.putIfAbsent(nextKey, value(nextKey));
+                String ePutAbs = testable.putIfAbsent(nextKey, value(nextKey));
+
+                assertEquals(oPutAbs, ePutAbs);
+                assertEquals(originMap.containsKey(nextKey), testable.containsKey(nextKey));
+            }
+            else {
+                String oRmv = originMap.remove(nextKey);
+                String eRmv = testable.remove(nextKey);
+
+                assertEquals(oRmv, eRmv);
+                assertEquals(originMap.get(nextKey), testable.get(nextKey));
+                assertEquals(originMap.containsKey(nextKey), testable.containsKey(nextKey));
+            }
+
+            Assert.assertEquals(originMap.size(), testable.size());
```

Comment:
It would be cool to iterate over these two maps and check that all elements are the same.